hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/index/RunIndexActionExecutor.java

```diff
@@ -0,0 +1,302 @@
+  // this needs to be volatile as it can be updated in the IndexingCheckTask spawned by this executor
+  // assumption is that only one indexer can execute at a time
+  private volatile String currentIndexedInstant;
+
+  private final TransactionManager txnManager;
+
+  public RunIndexActionExecutor(HoodieEngineContext context, HoodieWriteConfig config, HoodieTable<T, I, K, O> table, String instantTime) {
+    super(context, config, table, instantTime);
+    this.txnManager = new TransactionManager(config, table.getMetaClient().getFs());
+  }
+
+  @Override
+  public Option<HoodieIndexCommitMetadata> execute() {
+    HoodieTimer indexTimer = new HoodieTimer();
+    indexTimer.startTimer();
+
+    // ensure lock provider configured
+    if (!config.getWriteConcurrencyMode().supportsOptimisticConcurrencyControl() || StringUtils.isNullOrEmpty(config.getLockProviderClass())) {
+      throw new HoodieIndexException(String.format("Need to set %s as %s and configure lock provider class",
+          WRITE_CONCURRENCY_MODE.key(), OPTIMISTIC_CONCURRENCY_CONTROL.name()));
+    }
+
+    HoodieInstant indexInstant = table.getActiveTimeline()
+        .filterPendingIndexTimeline()
+        .filter(instant -> instant.getTimestamp().equals(instantTime) && REQUESTED.equals(instant.getState()))
+        .lastInstant()
+        .orElseThrow(() -> new HoodieIndexException(String.format("No requested index instant found: %s", instantTime)));
+    try {
+      // read HoodieIndexPlan
+      HoodieIndexPlan indexPlan = TimelineMetadataUtils.deserializeIndexPlan(table.getActiveTimeline().readIndexPlanAsBytes(indexInstant).get());
+      List<HoodieIndexPartitionInfo> indexPartitionInfos = indexPlan.getIndexPartitionInfos();
+      if (indexPartitionInfos == null || indexPartitionInfos.isEmpty()) {
+        throw new HoodieIndexException(String.format("No partitions to index for instant: %s", instantTime));
+      }
+      // transition requested indexInstant to inflight
+      table.getActiveTimeline().transitionIndexRequestedToInflight(indexInstant, Option.empty());
+      // start indexing for each partition
+      HoodieTableMetadataWriter metadataWriter = table.getMetadataWriter(instantTime)
+          .orElseThrow(() -> new HoodieIndexException(String.format("Could not get metadata writer to run index action for instant: %s", instantTime)));
+      // this will only build index upto base instant as generated by the plan, we will be doing catchup later
+      LOG.info("Starting Index Building");
+      metadataWriter.buildIndex(context, indexPartitionInfos);
+
+      // get all instants since the plan completed (both from active timeline and archived timeline)
+      // assumption is that all metadata partitions had same instant upto which they were scheduled to be indexed
+      table.getMetaClient().reloadActiveTimeline();
+      String indexUptoInstant = indexPartitionInfos.get(0).getIndexUptoInstant();
```

Comment:
should the uptoInstant be in a higher level struct then, to avoid `get(0)`

Reply:
i kept it at a lower level to have more flexibility so that we can support different indexUptoInstant for different indexes in future.